serverless/internal/storage/webstorage/fs.go

```diff
@@ -0,0 +1,290 @@
+		if err := createExclusive(leafFQ, []byte(strconv.FormatUint(seq, 16))); err != nil {
+			return 0, fmt.Errorf("couldn't create temporary leafhash file: %w", err)
+		}
+		// All done!
+		return seq, nil
+	}
+}
+
+// ScanSequenced calls the provided function once for each contiguous entry
+// in storage starting at begin.
+// The scan will abort if the function returns an error, otherwise it will
+// return the number of sequenced entries.
+func (fs *Storage) ScanSequenced(begin uint64, f func(seq uint64, entry []byte) error) (uint64, error) {
+	end := begin
+	for {
+		sp := filepath.Join(layout.SeqPath(fs.root, end))
+		entry, err := get(sp)
+		if errors.Is(err, os.ErrNotExist) {
+			// we're done.
+			return end - begin, nil
+		} else if err != nil {
+			return end - begin, fmt.Errorf("failed to read leafdata at index %d: %w", begin, err)
+		}
+		if err := f(end, entry); err != nil {
+			return end - begin, err
+		}
+		end++
+	}
+}
+
+// GetTile returns the tile at the given tile-level and tile-index.
+// If no complete tile exists at that location, it will attempt to find a
+// partial tile for the given tree size at that location.
+func (fs *Storage) GetTile(level, index, logSize uint64) (*api.Tile, error) {
+	tileSize := layout.PartialTileSize(level, index, logSize)
+	p := filepath.Join(layout.TilePath(fs.root, level, index, tileSize))
+	t, err := get(p)
+	if err != nil {
+		if !errors.Is(err, os.ErrNotExist) {
+			return nil, fmt.Errorf("failed to read tile at %q: %w", p, err)
+		}
+		return nil, err
+	}
+
+	var tile api.Tile
+	if err := tile.UnmarshalText(t); err != nil {
+		return nil, fmt.Errorf("failed to parse tile: %w", err)
+	}
+	return &tile, nil
+}
+
+// StoreTile writes a tile out to disk.
+// Fully populated tiles are stored at the path corresponding to the level &
+// index parameters, partially populated (i.e. right-hand edge) tiles are
+// stored with a .xx suffix where xx is the number of "tile leaves" in hex.
+func (fs *Storage) StoreTile(level, index uint64, tile *api.Tile) error {
+	tileSize := uint64(tile.NumLeaves)
+	glog.V(2).Infof("StoreTile: level %d index %x ts: %x", level, index, tileSize)
+	if tileSize == 0 || tileSize > 256 {
+		return fmt.Errorf("tileSize %d must be > 0 and <= 256", tileSize)
+	}
+	t, err := tile.MarshalText()
+	if err != nil {
+		return fmt.Errorf("failed to marshal tile: %w", err)
+	}
+
+	tDir, tFile := layout.TilePath(fs.root, level, index, tileSize%256)
+	tPath := filepath.Join(tDir, tFile)
+
+	if err := createExclusive(tPath, t); err != nil {
+		return fmt.Errorf("failed to write temporary tile file: %w", err)
+	}
+
+	return nil
+}
+
+// WriteCheckpoint stores a raw log checkpoint on disk.
+func (fs Storage) WriteCheckpoint(newCPRaw []byte) error {
+	oPath := filepath.Join(fs.root, layout.CheckpointPath)
+	return set(oPath, newCPRaw)
+}
+
+// ReadCheckpoint reads and returns the contents of the log checkpoint file.
+func ReadCheckpoint(root string) ([]byte, error) {
```

Comment:
Can we make this internal, and access this via `Storage.Checkpoint()` ? This is the same suggestion as that on `Load`, but a cleanup afterwards.

Reply:
Sadly not, for the same reason.